common/ssl_util.c

```diff
@@ -1090,6 +1089,210 @@ ssl_verify_signature(const char *cert_file, const char *signature_file, const ch
 	return ret;
 }
 
+int
+ssl_verify_signature_from_buf(const char *cert_buf, const uint8_t *sig_buf, size_t sig_len,
+			      const uint8_t *buf, size_t buf_len)
+{
+	ASSERT(cert_buf);
+	ASSERT(sig_buf);
+	ASSERT(buf);
+
+	int ret = 0;
+
+	// certificate variables
+	X509 *cert;
+	EVP_PKEY *key = NULL;
+	BIO *mem;
+
+	// signature variables
+	const EVP_MD *hash_fct;
+	EVP_MD_CTX *md_ctx = NULL;
+
+	// load certificate
+	mem = BIO_new(BIO_s_mem());
+	BIO_puts(mem, cert_buf);
+	cert = PEM_read_bio_X509(mem, NULL, 0, NULL);
+	BIO_free(mem);
+
+	const X509_ALGOR *sig_alg = X509_get0_tbs_sigalg(cert);
+	if (!sig_alg) {
+		ERROR("Error in signature verification (Failed to parse hash-algorithm)");
+		ret = -2;
+		goto error;
+	}
+
+	const char *hash_algo = asn1_object_to_hash_algo(sig_alg->algorithm);
+	if (!hash_algo) {
+		ERROR("Error in signature verification (Unsupported hash function)");
+		ret = -2;
+		goto error;
+	}
+
+	// load public key, digest, and verify signature
+	if ((key = X509_get_pubkey(cert)) == NULL) {
+		ERROR("Error in signature verification (loading pubkey failed)");
+		ret = -2;
+		goto error;
+	}
+
+	if ((hash_fct = EVP_get_digestbyname(hash_algo)) == NULL) {
+		ERROR("Error in signature verification (unable to initialize hash function)");
+		ret = -2;
+		goto error;
+	}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+	EVP_MD_CTX _md_ctx;
+	md_ctx = &_md_ctx;
+	EVP_MD_CTX_init(md_ctx);
+#else
+	if ((md_ctx = EVP_MD_CTX_new()) == NULL) {
+		ERROR("Allocating EVP_MD failed!");
+		goto error;
+	}
+#endif
+	EVP_VerifyInit(md_ctx, hash_fct);
+
+	if (!EVP_VerifyUpdate(md_ctx, buf, buf_len)) {
+		ERROR("Error in signature verification (reading/hashing signed file failed");
+		ret = -2;
+		goto error;
+	}
+
+	DEBUG("File hash computed to verify signature");
+
+	ret = EVP_VerifyFinal(md_ctx, sig_buf, sig_len, key);
+	if (ret != 1) {
+		DEBUG("Signature verification error");
+		// any error
+		if (ret == -1)
```

Comment:
thus does not make sense, the above expression check if (ret !=1)!

Reply:
Done as discussed

Reply:
sorry may bad! I read if "(ret !=1) if(ret==1)" missed the '-'